conans/client/loader.py

```diff
@@ -95,11 +96,18 @@ def _initialize_conanfile(conanfile, processed_profile):
         # Prepare the settings for the loaded conanfile
         # Mixing the global settings with the specified for that name if exist
         tmp_settings = processed_profile.processed_settings.copy()
-        if (processed_profile.package_settings_values and
-                conanfile.name in processed_profile.package_settings_values):
-            # Update the values, keeping old ones (confusing assign)
-            values_tuple = processed_profile.package_settings_values[conanfile.name]
-            tmp_settings.values = Values.from_list(values_tuple)
+        package_settings_values = processed_profile.package_settings_values
+        if package_settings_values:
+            pkg_settings = package_settings_values.get(conanfile.name)
+            if pkg_settings is None:
+                ref = "%s/%s@%s/%s" % (conanfile.name, conanfile.version,
+                                       conanfile._conan_user, conanfile._conan_channel)
+                for pattern, settings in package_settings_values.items():
+                    if fnmatch.fnmatchcase(ref, pattern):
+                        pkg_settings = settings
+                        break
+            if pkg_settings:
+                tmp_settings.values = Values.from_list(pkg_settings)
```

Comment:
Should't this code belong to the Profile class? `processed_profile.get_settings_for(reference)`